File: src/main/java/me/itzisonn_/meazy/parser/pasing_function/expression/PrimaryExpressionParsingFunction.java
package me.itzisonn_.meazy.parser.pasing_function.expression;

import me.itzisonn_.meazy.MeazyMain;
import me.itzisonn_.meazy.lexer.TokenTypes;
import me.itzisonn_.meazy.parser.ParsingContext;
import me.itzisonn_.meazy.text.TextKt;
import me.itzisonn_.meazy.lexer.Token;
import me.itzisonn_.meazy.lexer.TokenType;
import me.itzisonn_.meazy.parser.Parser;
import me.itzisonn_.meazy.parser.ast.expression.Expression;
import me.itzisonn_.meazy.parser.ast.expression.literal.*;
import me.itzisonn_.meazy.parser.InvalidStatementException;
import me.itzisonn_.meazy.parser.ast.expression.identifier.ClassIdentifier;
import me.itzisonn_.meazy.parser.ast.expression.identifier.FunctionIdentifier;
import me.itzisonn_.meazy.parser.ast.expression.identifier.VariableIdentifier;
import me.itzisonn_.meazy.parser.pasing_function.AbstractParsingFunction;
import me.itzisonn_.meazy.parser.pasing_function.ParsingHelper;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

@NullMarked
public class PrimaryExpressionParsingFunction extends AbstractParsingFunction<Expression> {
    public PrimaryExpressionParsingFunction() {
        super("primary_expression");
    }

    @Override
    public Expression parse(ParsingContext context, @Nullable Object... extra) {
        Parser parser = context.getParser();
        Token token = parser.getCurrent();
        TokenType tokenType = token.getType();

        if (tokenType.equals(TokenTypes.ID())) {
            if (parser.getTokens().size() > parser.getPos() + 1 && parser.getTokens().get(parser.getPos() + 1).getType().equals(TokenTypes.LEFT_PARENTHESIS())) {
                String id = parser.getCurrentAndNext().getValue();
                if (Character.isUpperCase(id.charAt(0))) return new ClassIdentifier(id);
                else return new FunctionIdentifier(id);
            }

            if (parser.getPos() > 0 && parser.getTokens().get(parser.getPos() - 1).getType().equals(TokenTypes.DOT())) {
                return new VariableIdentifier(parser.getCurrentAndNext().getValue());
            }

            String id = parser.getCurrentAndNext().getValue();
            if (Character.isUpperCase(id.charAt(0))) return new ClassIdentifier(id);
            else return new VariableIdentifier(id);
        }
        if (tokenType.equals(TokenTypes.NULL())) {
            parser.getCurrentAndNext();
            return new NullLiteral();
        }
        if (tokenType.equals(TokenTypes.NUMBER())) return new NumberLiteral(parser.getCurrentAndNext().getValue());
        if (tokenType.equals(TokenTypes.STRING())) return new StringLiteral(ParsingHelper.parseString(context));
        if (tokenType.equals(TokenTypes.BOOLEAN())) return new BooleanLiteral(Boolean.parseBoolean(parser.getCurrentAndNext().getValue()));
        if (tokenType.equals(TokenTypes.THIS())) {
            parser.getCurrentAndNext();
            return new ThisLiteral();
        }
        if (tokenType.equals(TokenTypes.LEFT_PARENTHESIS())) {
            parser.getCurrentAndNext();
            Expression value = parser.parse(MeazyMain.getDefaultIdentifier("expression"), Expression.class);
            parser.getCurrentAndNext(TokenTypes.RIGHT_PARENTHESIS(), TextKt.translatable("meazy:parser.expected", "right_parenthesis"));
            return value;
        }

        throw new InvalidStatementException(token.getLine(), TextKt.translatable("meazy:parser.exception.cant_parse", tokenType.getId()));
    }
}

File: src/main/java/me/itzisonn_/meazy/parser/pasing_function/expression/PrimaryExpressionParsingFunction.kt
package me.itzisonn_.meazy.parser.pasing_function.expression

import me.itzisonn_.meazy.lexer.TokenTypes.id
import me.itzisonn_.meazy.lexer.TokenTypes.boolean
import me.itzisonn_.meazy.lexer.TokenTypes.dot
import me.itzisonn_.meazy.lexer.TokenTypes.leftParenthesis
import me.itzisonn_.meazy.lexer.TokenTypes.`null`
import me.itzisonn_.meazy.lexer.TokenTypes.number
import me.itzisonn_.meazy.lexer.TokenTypes.rightParenthesis
import me.itzisonn_.meazy.lexer.TokenTypes.string
import me.itzisonn_.meazy.lexer.TokenTypes.`this`
import me.itzisonn_.meazy.parser.InvalidStatementException
import me.itzisonn_.meazy.parser.Parser
import me.itzisonn_.meazy.parser.ast.expression.Expression
import me.itzisonn_.meazy.parser.ast.expression.identifier.ClassIdentifier
import me.itzisonn_.meazy.parser.ast.expression.identifier.FunctionIdentifier
import me.itzisonn_.meazy.parser.ast.expression.identifier.VariableIdentifier
import me.itzisonn_.meazy.parser.ast.expression.literal.*
import me.itzisonn_.meazy.parser.pasing_function.ParsingFunction
import me.itzisonn_.meazy.parser.pasing_function.parseString
import me.itzisonn_.meazy.text.translatable

object PrimaryExpressionParsingFunction : ParsingFunction<Expression> {
    override fun Parser.parse(vararg extra: Any?): Expression {
        val token = current
        val tokenType = token.type

        if (tokenType == id) {
            if (size > pos + 1 && this[pos + 1].type == leftParenthesis) {
                val id = consume().value
                return if (Character.isUpperCase(id[0])) ClassIdentifier(id)
                else FunctionIdentifier(id)
            }

            if (pos > 0 && this[pos - 1].type == dot) {
                return VariableIdentifier(consume().value)
            }

            val id = consume().value
            return if (Character.isUpperCase(id[0])) ClassIdentifier(id) else VariableIdentifier(id)
        }
        if (tokenType == `null`) {
            consume()
            return NullLiteral()
        }
        if (tokenType == number) return NumberLiteral(consume().value)
        if (tokenType == string) return StringLiteral(parseString())
        if (tokenType == boolean) return BooleanLiteral(consume().value.toBoolean())
        if (tokenType == `this`) {
            consume()
            return ThisLiteral()
        }
        if (tokenType == leftParenthesis) {
            consume()
            val value = parse(ExpressionParsingFunction)
            consume(rightParenthesis, translatable("meazy:parser.expected", "right_parenthesis"))
            return value
        }

        throw InvalidStatementException(token.line, translatable("meazy:parser.exception.cant_parse", tokenType.id))
    }
}
